src/python/pants/build_graph/address.py

```diff
@@ -171,11 +179,18 @@ def file_to_address(self) -> "Address":
 
         expected_prefix = f"..{os.path.sep}" * parent_count
         if self.target_component[: self.target_component.rfind(os.path.sep) + 1] != expected_prefix:
-            raise InvalidTargetName(
-                "A target may only be defined in a directory containing a file that it owns in "
-                f"the filesystem: `{self.target_component}` is not at-or-above the file "
-                f"`{self.path_component}`."
-            )
+            if sum(1 for c in os.path.split(self.target_component) if c != "..") > 1:
+                raise InvalidTargetNameWithSlash(
+                    "The target name `{self.target_component}` is ambiguous, because it might "
+                    "represent either a target with a slash in the name, or a target nested inside "
+                    "of a directory."
```

Comment:
It'd be good to add a remedy:

> Please rename the `name=` field for the target at {path_component} to not use slashes.

Reply:
Note also that this error doesn't mention the path_component atm, so it's not clear where the issue comes from.